Hotfix/p9 quarter

maddyscientist opened the pull request:
This is a bugfix PR
* Quarter precision is defined using the `int8_t` and not `char`.  This removes the ambiguity of the sign, and fixes quarter precision which was broken when QUDA when compiled on ARM and POWER platforms
* Deployed `instantiate` helper in more places to remove explicit template instantiation
* Moved copy_gauge.cu and copy_color_spinor.cu files to cpp since they contain no CUDA
* Improved performance of `computeCoarseClover` which was implemented inefficiently (@weinbe2 request)

I have done a full ctest run for all dslash types with quarter precision enabled, and all works fine on P9.  Also tested with Jitify as well, and things are good there.

maddyscientist committed:
Global replacement of char type with int8_t for quarter precision - this is to fix quarter precision on POWER/ARM platforms.  As part of this work, replaced if/else precision template instantiations where possible

maddyscientist committed:
Missing char->int8_t from prior commit

maddyscientist committed:
Remove some pre-Kepler guards that are no longer needed

maddyscientist committed:
Move copy_gauge.cu and copy_color_spinor.cu to copy_gauge.cpp and copy_color_spinor.cpp, respectively

maddyscientist committed:
Improve performance of computeCoarseClover kernel: split triple matrix product into two steps to prevent redundant computation

weinbe2 commented on include/instantiate.h:
```diff
@@ -232,7 +232,118 @@ namespace quda
+  constexpr void instantiatePrecision(F &field, Args &&... args)
+  {
+    if (field.Precision() == QUDA_DOUBLE_PRECISION) {
+      // always instantiate double precision
```
Why is double always being instantiated?

maddyscientist replied:
This is a special case template: for the interface we need to always instantiate double precision for the copy in/out kernels.

weinbe2 replied:
Understood, please add that to the doxygen then

weinbe2 commented on include/instantiate.h:
```diff
@@ -232,7 +232,118 @@ namespace quda
+  }
+
+  /**
+     @brief The instantiatePrecision function is used to instantiate
```
Why do we need two `instantiatePrecision` functions (different template signatures noted)? If having two is unavoidable, is there a better naming option than `instantiatePrecision` and `instantiatePrecision2`?

maddyscientist replied:
Cleaning this up is easy with C++17 😜   This solution was expedient in the meantime....  Open to naming suggestions.  This is really just a placeholder.

weinbe2 replied:
Fair, just please make the use for this version vs the other one clearer from the doxygen

maddyscientist replied:
Will do.

maddyscientist committed:
Merge branch 'develop' of github.com:lattice/quda into hotfix/p9-quarter

weinbe2 commented:
Open question, do we want to go down the rabbit hole/hell of replacing `short` with `int16_t` for aesthetic/consistency reasons?

maddyscientist commented:
I think there’s no rush to do this now.  Eventually, once we have `instantiate` everywhere, the number of shorts and int8_t actually in quda will be limited to just a few headers.  At that point, sure, but not now.

weinbe2 commented on lib/copy_clover.cu:
```diff
@@ -129,111 +128,67 @@ namespace quda {
 #ifdef GPU_CLOVER_DIRAC
-    if (out.Precision() == QUDA_HALF_PRECISION && out.Order() > 4) 
+    if (out.Precision() < QUDA_SINGLE_PRECISION && out.Order() > 4) 
       errorQuda("Half precision not supported for order %d", out.Order());
```
Perhaps change this to say "Fixed precision..." (here and below)

hummingtree reviewed (approved):
The only request is maybe we can have something like a `instantiateGaugeOrder` template available, similar to `instantiatePrecision`.

maddyscientist committed:
Address review comments

maddyscientist committed:
Remove erroneous check against using half precision with Laplace operator, and erroneous copy in blas_test

maddyscientist committed:
Update to the latest version of Jitify.  This fixes POWER9 + Jitify support (closes #801).

maddyscientist committed:
Added workaround for flaky CUDA 9.x behaviour

maddyscientist committed:
Fix compiler warning contract.cu

maddyscientist commented:
I have pushed some updates, so we should be good with this PR now
* Addressed @weinbe2's review comments
* Updated to the latest version of Jitify - this closes an old bug #801 (Jitify didn't work on POWER9)
* Added a workaround for CUDA 9.x Jenkins failure 

@hummingtree to address your question: `instantiateGaugeOrder` has limited use since it would only be in the copy_gauge routines.  In most places we simply infer the gauge order for native fields from the precision and reconstruct values.

weinbe2 reviewed (approved):
I've given this a few iterations of visual reviews, plus asked @maddyscientist to do a few due-diligence tests related to checking jitify+MG, etc, and all looks good. Assuming nothing weird turns up during the current set of Jenkins tests this will be good to merge.
